// spnk-core/src/Platforms/YouTube/DownloadVideo.js
"use strict";
const {DownloadVideo} = require('../../Global/FFMPEG.js');
const {getData} = require('./getInfo.js');
//Options - {format: "mp4", quality: "1080p"}
const D = (url, options = {}) => {
    getData(url).then(video => {
        if (video.isLive) return console.log('Live video is not a support');

        if (!options.format) options = {...options, format: "mp4"};
        if (!options.name) options = {...options, name: video.id};
        let format = FindFormats(video.formats, options.quality);

        return DownloadVideo(format, options);
    })
}
module.exports.DownloadVideo = D;

function FindFormats(formats, quality) {
    let videoUrl = FindVideo(formats, quality);
    let audioUrl = FindAudio(formats);
    return {
        video: videoUrl.url,
        audio: audioUrl.url
    }
}
function FindVideo(formats, quality = null) {
    let format;
    if (quality) {
        format = formats.find(f => f.video.quality === quality);
    } else {
        format = formats.find(f => f.playable === 'VideoOnly');
    }
    return format;
}
function FindAudio(formats) {
    let audioUrl = formats.filter(f => f.playable && f.audio.codec === 'opus').sort((a, b) => (b.audio.SampleRate - a.audio.SampleRate));
    return audioUrl[0];
}
