// spnk-core/src/Platforms/YouTube/getChannelInfo.js
"use strict";
const request = require('../../Global/Request.js');

class getChannelInfo {
    callback = (base) => {
        return new Promise((res) => {
            request.makeRequest(`https://www.youtube.com/channel/${base.channelId || base.channelID}/channels?flow=grid&view=0&pbj=1`).then(response => response.json()).then(channel => {
                let data = channel[1]?.response || null;

                if (!data) return res({
                    id: base.channelId || base.channelID,
                    title: base.author || base.name,
                    url: `https://www.youtube.com/channel/${base.channelId || base.channelID}`,
                    thumbnails: {
                        url: null,
                        width: 0,
                        height: 0
                    },
                    isVerified: undefined
                })

                let channelMetaData = data.metadata.channelMetadataRenderer, channelHeaderData = data.header.c4TabbedHeaderRenderer;
                return res({
                    id: channelMetaData.externalId,
                    title: channelMetaData.title,
                    url: channelMetaData.vanityChannelUrl,
                    thumbnails: channelHeaderData.avatar.thumbnails[2] ?? channelHeaderData.avatar.thumbnails[1] ?? channelHeaderData.avatar.thumbnails[0],
                    isVerified: !!channelHeaderData.badges?.find((badge) => badge.metadataBadgeRenderer.tooltip === 'Verified' || badge.metadataBadgeRenderer.tooltip === 'Official Artist Channel'),
                });
            })
        });
    }
}
module.exports = new getChannelInfo().callback;
